### PR TITLE
The raster correction function and tests

### PR DESCRIPTION
**Pull request by ecobost**
Copied the version from Matlab (commons.ne7.ip.correctRaster) and checked against it.

### CODE REVIEW

File: python/pipeline/utils/galvo_corrections.py [hunk 1]
```diff
@@ -20,32 +22,68 @@ def correct_motion(img, xymotion):
...
-    """
-    raster correction for resonant scanners.
-    :param img: 5D image [x, y, nchannel, nslice, nframe].
-    :param raster_phase: phase difference beetween odd and even lines.
-    :param fill_fraction: ratio between active acquisition and total length of the scan line. see scanimage.
-    :return: raster-corrected image [x, y, nchannel, nslice, nframe].
+def correct_raster(scan, raster_phase, fill_fraction, in_place=True):
+    """ Raster correction for resonant scanners.
+
+    Corrects EM images in n-dimensional scans, usual shape is [image_height, image_width,
```
Comment (fabiansinz) [inline python/pipeline/utils/galvo_corrections.py]: It's not EM, it is 2-Photon images.

Comment (fabiansinz) [inline python/pipeline/utils/galvo_corrections.py]: Let's make the shape required (i.e. replace usual by required.)

File: python/pipeline/utils/galvo_corrections.py [hunk 2]
```diff
@@ -59,20 +97,20 @@ def plot_raster(filename, key):
     from pipeline import preprocess, experiment
     from tiffreader import TIFFReader
     reader = TIFFReader(filename)
-    img=reader[:, :, 0, 0, 100]
+    img = reader[:, :, 0, 0, 100]
     raster_phase = (preprocess.Prepare.Galvo() & key).fetch1['raster_phase']
     newim = correct_raster(img, raster_phase, reader.fill_fraction)
     nnewim = correct_raster(newim, -raster_phase, reader.fill_fraction)
     print(np.mean(img - nnewim))
```
Comment (fabiansinz) [inline python/pipeline/utils/galvo_corrections.py]: Can you remove that print?

File: python/pipeline/utils/galvo_corrections.py [hunk 3]
```diff
@@ -59,20 +97,20 @@ def plot_raster(filename, key):
...
     with sns.axes_style('white'):
-        fig=plt.figure(figsize=(15,8))
-        gs=plt.GridSpec(1,3)
-        ax1=fig.add_subplot(gs[0,0])
-        ax1.imshow(img[:,:,0,0,0], cmap=plt.cm.gray)
-        ax2=fig.add_subplot(gs[0,1])
-        ax2.imshow(newim[:,:,0,0,0],cmap=plt.cm.gray)
-        ax3=fig.add_subplot(gs[0,2])
-        ax3.imshow(nnewim[:,:,0,0,0], cmap=plt.cm.gray)
+        fig = plt.figure(figsize=(15, 8))
```
Comment (fabiansinz) [inline python/pipeline/utils/galvo_corrections.py]: That can be shorter with `fig, ax = plt.subplots(1,3)`.

File: python/pipeline/utils/galvo_corrections.py [hunk 4]
```diff
@@ -59,20 +97,20 @@ def plot_raster(filename, key):
     from pipeline import preprocess, experiment
```
Comment (fabiansinz) [inline python/pipeline/utils/galvo_corrections.py]: Can you move the matplotlib and seaborn import into the plot function? I'd prefer to import it only when we plot and not by default.

File: python/tests/test_preprocess.py [hunk 5]
```diff
@@ -17,19 +14,54 @@ def test_motioncorrect():
...
+
+    #int to double
+    test_scan = np.arange(128).reshape([4, 4, 2, 2, 2])
+    result = galvo_corrections.correct_raster(test_scan, raster_phase=0.5, fill_fraction=1)
+    assert(result.dtype == np.double)
+
+def test_raster_correction_with_smaller_input():
+    test_scan = np.double(np.arange(32).reshape([4, 4, 2]))
+    result = galvo_corrections.correct_raster(test_scan, raster_phase=0.5, fill_fraction=1)
+    assert (result.ndim == 3)
```
Comment (fabiansinz) [inline python/tests/test_preprocess.py]: You could also test whether it returns the exact same shape.

**Comment by ecobost** [comment]
Added the motion correction function and tests and made the changes for raster correction. Both passed all test so they are probably alright

File: python/pipeline/utils/galvo_corrections.py [hunk 6]
```diff
@@ -20,32 +22,68 @@ def correct_motion(img, xymotion):
...
+        raise PipelineException('Scan needs to be a numpy array.')
+    if scan.ndim < 2:
+        raise PipelineException('Scan with less than 2 dimensions.')
+    if not np.issubdtype(scan.dtype, np.float):
+        print('Changing scan from', str(scan.dtype), 'to double')
+        scan = np.double(scan)
+
+    # Get some dimensions
+    original_shape = scan.shape
+    image_height = original_shape[0]
```
Comment (dimitri-yatsenko) [inline python/pipeline/utils/galvo_corrections.py]: image_height, image_width = original_shape

File: python/pipeline/utils/galvo_corrections.py [hunk 7]
```diff
@@ -20,32 +22,68 @@ def correct_motion(img, xymotion):
...
+    image_width = original_shape[1]
+    half_width = image_width / 2
+
+    # ??
```
Comment (dimitri-yatsenko) [inline python/pipeline/utils/galvo_corrections.py]: comment should say something like "create interpolation points for sinusoidal raster"

**Pull request merged**